### ISSUE TITLE
Add weight replica load-balance algorithm.

### ISSUE THREAD

**Issue by cherubimLV** [apache/shardingsphere#13933]
In the process of using ShardingSphere-JDBC, it is found that the important weight polling of the polling algorithm of the reading library has not been implemented, but in fact, the reading library is often accessed according to the weight during project operation to prevent the difference in data service performance configuration from causing the project to stability.

**Comment by cherubimLV** [comment]
```java
@Getter
@Setter
public class WeightReplicaLoadBalanceAlgorithm implements ReplicaLoadBalanceAlgorithm {
    private static final ConcurrentHashMap<String, Double[]> WEIGHT_MAP = new ConcurrentHashMap<>();

    private Properties props = new Properties();

    @Override
    public Properties getProps() {
        return this.props;
    }

    @Override
    public void setProps(final Properties props) {
        this.props = props;
    }

    @Override
    public String getType() {
        return "WEIGHT";
    }

    @Override
    public String getDataSource(final String name, final String writeDataSourceName, final List<String> readDataSourceNames) {
        Double[] weight = WEIGHT_MAP.containsKey(name) ? WEIGHT_MAP.get(name) : initWeight(readDataSourceNames);
        WEIGHT_MAP.putIfAbsent(name, weight);
        String dataSourceName = getDataSourceName(readDataSourceNames, weight);
        return dataSourceName;
    }

    private String getDataSourceName(final List<String> readDataSourceNames, final Double[] weight) {
        double randomWeight = ThreadLocalRandom.current().nextDouble(0, 1);
        int index = Arrays.binarySearch(weight, randomWeight);
        if (index < 0) {
            index = -index - 1;
            return index >= 0 && index < weight.length && randomWeight < weight[index] ? readDataSourceNames.get(index) : readDataSourceNames.get(readDataSourceNames.size() - 1);
        } else {
            return readDataSourceNames.get(index);
        }
    }

    private Double[] initWeight(final List<String> readDataSourceNames) {
        Double[] weights;
        Double[] exactWeights = new Double[readDataSourceNames.size()];
        Integer index = 0;
        Double sum = 0D;
        for (String readDataSourceName : readDataSourceNames) {
            Object weightObject = props.get(readDataSourceName);
            if (weightObject == null) {
                throw new IllegalStateException("Read database access weight is not configured：" + readDataSourceName);
            }
            Double weight;
            try {
                weight = Double.valueOf(weightObject.toString());
            } catch (NumberFormatException e) {
                throw new NumberFormatException("Read database weight configuration error, configuration parameters:" + weightObject.toString());
            }
            if (Double.isInfinite(weight)) {
                weight = 10000.0D;
            }
            if (Double.isNaN(weight)) {
                weight = 1.0D;
            }
            exactWeights[index++] = weight;
            sum += weight;
        }
        for (int i = 0; i < index; i++) {
            if (exactWeights[i] <= 0) {
                continue;
            }
            exactWeights[i] = exactWeights[i] / sum;
        }
        weights = new Double[readDataSourceNames.size()];
        double randomRange = 0D;
        for (int i = 0; i < index; i++) {
            weights[i] = randomRange + exactWeights[i];
            randomRange += exactWeights[i];
        }
        if (index != 0 && Math.abs(weights[index - 1] - 1.0D) >= 0.0001) {
            throw new IllegalStateException("The cumulative weight is calculated incorrectly, and the sum of the probabilities is not equal to 1.");
        }
        return weights;
    }
}
```

**Comment by cherubimLV** [comment]
```yml
spring:
	shardingsphere:
		rules:
          readwrite-splitting:
            data-sources:
              dataSources:
                write-data-source-name: master
                read-data-source-names: slave1,slave2
                load-balancer-name: weight
            load-balancers:
                weight:
                  type: WEIGHT  
                  props:
                    slave1: 1
                    slave2: 1

```

**Comment by flycash** [comment]
do you guys need help? I can help to do this

**Comment by RaigorJiang** [comment]
Hi @cherubimLV @flycash 
This idea looks good, you can try to submit a PR, thanks~

**Comment by cherubimLV** [comment]
Hi @RaigorJiang 
I have already submit  a PR

**Comment by RaigorJiang** [comment]
Fixed by #13943.
